pkg/services/ngalert/notifier/channels/email.go

```diff
@@ -121,6 +124,39 @@ func (en *EmailNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool,
 		},
 	}
 
+	// TODO: modify the email sender code to support multiple file or image URL
+	// fields. We cannot use images from every alert yet.
+	imgToken := getTokenFromAnnotations(as[0].Annotations)
+	if len(imgToken) != 0 {
+		timeoutCtx, cancel := context.WithTimeout(ctx, ImageStoreTimeout)
+		imgURL, err := en.images.GetURL(timeoutCtx, imgToken)
+		cancel()
+		if err != nil {
+			if !errors.Is(err, ErrImagesUnavailable) {
+				// Ignore errors. Don't log "ImageUnavailable", which means the storage doesn't exist.
+				en.log.Warn("failed to retrieve image url from store", "error", err)
+			}
+		} else if len(imgURL) > 0 {
+			cmd.Data["ImageLink"] = imgURL
+		} else { // Try to upload
+			timeoutCtx, cancel := context.WithTimeout(ctx, ImageStoreTimeout)
+			imgPath, err := en.images.GetFilepath(timeoutCtx, imgToken)
+			cancel()
+			if err != nil {
+				if !errors.Is(err, ErrImagesUnavailable) {
+					// Ignore errors. Don't log "ImageUnavailable", which means the storage doesn't exist.
+					en.log.Warn("failed to retrieve image url from store", "error", err)
+				}
+			} else if len(imgPath) != 0 {
+				file, err := os.Stat(imgPath)
+				if err == nil {
+					cmd.EmbeddedFiles = []string{imgPath}
+					cmd.Data["EmbeddedImage"] = file.Name()
+				}
```

Comment:
Should we log here if err != nil?

Reply:
Yeah. I think the most common case for this would be old data that gets cleaned off the filesystem. Good call.